src/components/structures/MessagePanel.js

```diff
@@ -286,6 +287,47 @@ module.exports = React.createClass({
 
             var last = (i == lastShownEventIndex);
 
+            var isMembershipChange = (e) =>
+                e.getType() === 'm.room.member'
+                && ['join', 'leave'].indexOf(e.event.content.membership) !== -1
+                && (!e.event.prev_content || e.event.content.membership  !== e.event.prev_content.membership);
+
+            // Wrap consecutive member events in a ListSummary
+            if (isMembershipChange(mxEv)) {
+                let summarisedEvents = [mxEv];
+                i++;
+                for (;i < this.props.events.length; i++) {
+                    let collapsedMxEv = this.props.events[i];
+
+                    if (!isMembershipChange(collapsedMxEv)) {
+                        i--;
+                        break;
+                    }
+                    summarisedEvents.push(collapsedMxEv);
+                }
+                let renderEvents = (pEvent, events) => {
+                    if (events.length === 0) {
+                        return null;
+                    }
+                    return events.map(
+                        (e) => {
+                            let ret = this._getTilesForEvent(pEvent, e);
+                            pEvent = e;
+                            return ret;
+                        }
+                    ).reduce((a,b) => a.concat(b));
+                };
+                ret.push(
+                    <MemberEventListSummary
+                        events={summarisedEvents}
+                        previousEvent={prevEvent}
+                        renderEvents={renderEvents}
```

Comment:
passing in a `render` callback feels pretty icky here.

An alternative might be to host the event filtering logic in `MemberEventListSummary.render` to here, then generate the tiles here, and then set them as children of the `MemberEventListSummary`:

```
const filteredEvents = filterEvents(summarisedEvents); /* as per MemberEventListSummary.render() */
const tiles = filteredEvents.map( /* as above*/ );
<MemberEventListSummary events={summarisedEvents}>
    {tiles}
</MemberEventListSummary>
```

and in `MemberEventListSummary.render()`:
```
expandedEvents = this.props.children;
```

WDYT?